app/code/Magento/InventoryCatalog/Model/LegacySynchronization/ToLegacyCatalogInventory/SetDataToLegacyInventory.php

```diff
@@ -0,0 +1,131 @@
+     * @var DefaultSourceProviderInterface
+     */
+    private $defaultSourceProvider;
+
+    /**
+     * SetDataToLegacyCatalogInventory constructor.
+     * @param DefaultSourceProviderInterface $defaultSourceProvider
+     * @param SetDataToLegacyStockItem $setDataToLegacyStockItem
+     * @param GetLegacyStockItemsByProductIds $getLegacyStockItemsByProductIds
+     * @param StockStateProviderInterface $stockStateProvider
+     * @param Processor $indexerProcessor
+     * @param GetProductIdsBySkus $getProductIdsBySkus
+     * @SuppressWarnings(PHPMD.LongVariable)
+     */
+    public function __construct(
+        DefaultSourceProviderInterface $defaultSourceProvider,
+        SetDataToLegacyStockItem $setDataToLegacyStockItem,
+        GetLegacyStockItemsByProductIds $getLegacyStockItemsByProductIds,
+        StockStateProviderInterface $stockStateProvider,
+        Processor $indexerProcessor,
+        GetProductIdsBySkus $getProductIdsBySkus
+    ) {
+        $this->setDataToLegacyStockItem = $setDataToLegacyStockItem;
+        $this->getLegacyStockItemsByProductIds = $getLegacyStockItemsByProductIds;
+        $this->stockStateProvider = $stockStateProvider;
+        $this->indexerProcessor = $indexerProcessor;
+        $this->getProductIdsBySkus = $getProductIdsBySkus;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+    }
+
+    /**
+     * Synchronously execute legacy alignment
+     *
+     * @param array $sourceItemsData
+     * @throws LocalizedException
+     */
+    public function execute(array $sourceItemsData): void
+    {
+        $productIds = [];
+        foreach ($sourceItemsData as $sourceItemData) {
+            $sku = (string) $sourceItemData[SourceItemInterface::SKU];
+
+            if ($sourceItemData[SourceItemInterface::SOURCE_CODE] !== $this->defaultSourceProvider->getCode()) {
+                throw new LocalizedException(__('Only default source can synchronize legacy stock'));
+            }
+
+            try {
+                $productId = (int) current($this->getProductIdsBySkus->execute([$sku]));
+            } catch (NoSuchEntityException $e) {
+                continue;
+            }
+
+            $legacyStockItems = $this->getLegacyStockItemsByProductIds->execute([$productId]);
+            if (!isset($legacyStockItems[$productId])) {
+                continue;
+            }
+
+            $legacyStockItem = $legacyStockItems[$productId];
+            $isInStock = (int) $sourceItemData[SourceItemInterface::STATUS];
+            $quantity = (float) $sourceItemData[SourceItemInterface::QUANTITY];
+
+            if ($legacyStockItem->getManageStock()) {
+                $legacyStockItem->setIsInStock($isInStock);
+                $legacyStockItem->setQty($quantity);
+
+                if (false === $this->stockStateProvider->verifyStock($legacyStockItem)) {
+                    $isInStock = 0;
+                }
+            }
+
+            $this->setDataToLegacyStockItem->execute(
```

Comment:
maybe make sense to change this service in a way to work with multiple entities as well?